paddle/framework/scope.h

```diff
@@ -0,0 +1,87 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "paddle/framework/variable.h"
+#include "paddle/platform/assert.h"
+
+namespace paddle {
+namespace framework {
+
+/**
+ * Scope is an association of a name to Variable. All variables belong to
+ * Scope. You need to specify a scope to run a Net, i.e., `net.Run(&scope)`.
+ * One net can run in different scopes and update different variable in the
+ * scope.
+ */
+class Scope {
+ private:
+  explicit Scope(const std::shared_ptr<Scope>& parent = nullptr)
+      : parent_(parent) {}
+
+ public:
+  static std::shared_ptr<Scope> Create(
+      const std::shared_ptr<Scope>& parent = nullptr) {
+    return std::make_shared<Scope>(Scope(parent));
```

Comment:
According to https://google.github.io/styleguide/cppguide.html#Default_Arguments, we shouldn't use default argument here.

My proposal in https://github.com/PaddlePaddle/Paddle/pull/2629#discussion_r124703915 seems OK.